src/metrics.cc

```diff
@@ -378,6 +412,156 @@ Metrics::PollCacheMetrics(std::shared_ptr<RequestResponseCache> response_cache)
+  }
+
+  std::string _;
+  std::istringstream iss(line);
+  // Use _ to skip "cpu" at start of line
+  if (!(iss >> _ >> info->user >> info->nice >> info->system >> info->idle >>
+        info->iowait >> info->irq >> info->softirq >> info->steal)) {
+    return Status(
+        Status::Code::INTERNAL,
+        "Failed to parse aggregate CPU info in /proc/stat.");
+  }
+  return Status::Success;
+#endif  // OS
+}
+
+Status
+Metrics::ParseMemInfo(std::shared_ptr<MemInfo> info)
+{
+#ifdef _WIN32
+  return Status(
+      Status::Code::INTERNAL, "Memory metrics not supported on Windows.");
+#else
+  std::ifstream ifs("/proc/meminfo");
+  if (!ifs.good()) {
+    return Status(Status::Code::INTERNAL, "Failed to open /proc/meminfo.");
+  }
+
+  std::string line;
+  constexpr uint64_t KB = 1024;
+  while (std::getline(ifs, line)) {
+    std::istringstream iss(line);
+    std::string name;
+    uint64_t value = 0;
+    if (iss >> name >> value) {
+      name.pop_back();
+      (*info)[name] = value * KB;
+    } else {
+      return Status(
+          Status::Code::INTERNAL, "Encountered error parsing /proc/meminfo.");
+    }
+  }
+
+  if (info->find("MemTotal") == info->end() ||
+      info->find("MemAvailable") == info->end()) {
+    return Status(
+        Status::Code::INTERNAL,
+        "Failed to find desired values in /proc/meminfo.");
+  }
+
+  // "Used" memory can be defined in many different ways. While many
+  // older applications consider "used = total - (free + cached)", a more
+  // accurate measure of available memory "MemAvailable" was added,
+  // so we choose "used = total - available" for a more accurate measure.
+  // This may change in the future if not sufficient for most use cases.
+  // See https://stackoverflow.com/a/35019697.
+  const uint64_t mem_total_bytes = (*info)["MemTotal"];
+  const uint64_t mem_avail_bytes = (*info)["MemAvailable"];
+  if (mem_total_bytes >= mem_avail_bytes) {
+    (*info)["MemUsed"] = mem_total_bytes - mem_avail_bytes;
+  } else {
+    return Status(
+        Status::Code::INTERNAL,
+        "Available bytes shouldn't be greater than Total bytes");
+  }
+
+  return Status::Success;
+#endif  // OS
+}
+
+double
+Metrics::CpuUtilization(
+    std::shared_ptr<CpuInfo> info_new, std::shared_ptr<CpuInfo> info_old)
+{
+  if (info_new == nullptr || info_old == nullptr) {
+    LOG_ERROR << "Received nullptr on computing CPU utilization.";
+    return 0;
+  }
+
+  // Account for overflow
+  const auto wrap_sub = [](uint64_t a, uint64_t b) {
+    return (a > b) ? (a - b) : 0;
+  };
+  uint64_t util_diff = wrap_sub(info_new->user, info_old->user) +
+                       wrap_sub(info_new->nice, info_old->nice) +
+                       wrap_sub(info_new->system, info_old->system) +
+                       wrap_sub(info_new->irq, info_old->irq) +
+                       wrap_sub(info_new->softirq, info_old->softirq) +
+                       wrap_sub(info_new->steal, info_old->steal);
```

Comment:
How to interpret "util" in this context? Can we sum "user" and "nice" with other fields?

Reply:
All fields other than `idle` and `iowait` are considered "active/utilized" time spent on CPU. Would you prefer a breakdown like [this](https://github.com/htop-dev/htop/blob/15652e7b8102e86b3405254405d8ee5d2a239004/linux/LinuxProcessList.c#L1286-L1294)?

Reply:
I just don’t have the knowledge of what the fields mean, I don’t have preference as long as the fields are in the same “unit”

Reply:
Ah yes they're all in the same [units](https://man7.org/linux/man-pages/man5/procfs.5.html):
> The amount of time, measured in units of USER_HZ
>                     (1/100ths of a second on most architectures ...